src/NServiceBus.Core/Transports/FileBased/DirectoryBasedTransaction.cs

```diff
@@ -0,0 +1,91 @@
+namespace NServiceBus.Transports.FileBased
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+
+    class DirectoryBasedTransaction
+    {
+        public DirectoryBasedTransaction(string basePath)
+        {
+            this.basePath = basePath;
+            var transactionId = Guid.NewGuid().ToString();
+
+            transactionDir = Path.Combine(basePath, ".pending", transactionId);
+            commitDir = Path.Combine(basePath, ".committed", transactionId);
+        }
+
+        public string FileToProcess { get; set; }
+
+        public void BeginTransaction(string incomingFilePath)
+        {
+            Directory.CreateDirectory(transactionDir);
+            FileToProcess = Path.Combine(transactionDir, Path.GetFileName(incomingFilePath));
+            File.Move(incomingFilePath, FileToProcess);
+        }
+
+        public void Commit()
+        {
+            var dispatchFile = Path.Combine(transactionDir, "dispatch.txt");
+            File.WriteAllLines(dispatchFile, outgoingFiles.Select(file => $"{file.TxPath}=>{file.TargetPath}").ToArray());
+
+            Directory.Move(transactionDir, commitDir);
+            committed = true;
+        }
+
+
+        public void Rollback()
+        {
+            //rollback by moving the file back to the main dir
+            File.Move(FileToProcess, Path.Combine(basePath, Path.GetFileName(FileToProcess)));
+            Directory.Delete(transactionDir, true);
+        }
+
+
+        public void Enlist(string messagePath, List<string> messageContents)
+        {
+            var txPath = Path.Combine(transactionDir, Path.GetFileName(messagePath));
```

Comment:
Now the audit message and incoming message are represented by the same file in transaction folder. Currently that works fine but I'm not sure if we should make such assumption. Maybe we could add destination to the .txt filename?